Issue 679 to redesign filter button by date

OlehFilyak opened the pull request:
Pull Request Template
 

### Type of Pull Request *
- [] CHANGE (fix or feature that would cause existing functionality to not work as expected)
- [X] FEATURE (non-breaking change which adds functionality)
- [ ] BUGFIX (non-breaking change which fixes an issue)
- [ ] ENHANCEMENT  (non-breaking change which improves existing functionality)
- [ ] NONE (if none of the other choices apply. For example, tooling, build system, CI, docs, etc.)

### Related links
 
**Issue link**
[#679 To redesign filter button by date ](https://github.com/ita-social-projects/dokazovi-requirements/issues/679)
 )
 
**Story link**
[#679 1 Story]( https://github.com/ita-social-projects/dokazovi-requirements/issues/679#issue-1871973783)
 
### Description *
At first, when you don't choose any date, there are no styles for today's day. When you choose someday styles add. If you choose the same day we will reset the styles and the value in the global state.

OlehFilyak committed:
feat(DatePicker): change styles for selectedDay

Initially, the DatePicker style for SelectedDay is not displayed when you open it. As soon as it is selected, the background color is only for the selected element.

OlehFilyak committed:
feat(DatePicker): change setting styles and add reseting Date for DatePicker

At first, when you don't choose any date, there are no styles for today's day. When you choose someday styles add. If you choose the same day we will reset the styles and the value in the global state.

OlehFilyak committed:
feat (DatePicker) add icon

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -32,18 +35,100 @@ interface IMaterialsDate {
+    firstSelectedDay: ISelectedDay;
+    secondSelectedDay: ISelectedDay;
+  },
+  action: { type: string; payload: ISelectedDay },
```
it always makes sense to type `type` part. then TS will save our backs when we make a typo in `type` value

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -32,18 +35,100 @@ interface IMaterialsDate {
+  secondSelectedDay: ISelectedDay;
+} => {
+  switch (action.type) {
+    case 'SET_SELECTED_DAY':
```
instead of using generic `SET_SELECTED_DAY` type, I propose to have separate `SET_START_DAY` and `SET_END_DAY` right from the start. This give us 2 benefits: 
1. better readability
a. (no `if .. ` nested into `switch/case` - the less nesting the easier to understand) 
b. explicit types for `START` and `END` instead of passing this as a part of payload
2, more flexibility: maybe it has not been stated yet, but typically it's expected that when we set end date before current start date or in opposite - setting start date after current end date, we either clear or update counterpart(e.g. when setting start date before current end date we either clear end date at all or update it to be equal to start date); even if this was not required yet, extra logic is still easier to apply with separate action types

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -32,18 +35,100 @@ interface IMaterialsDate {
   setChanges,
 }) => {
+  const [datePickerState, dispatch] = useReducer(datePickerReducer, {
+    firstSelectedDay: { position: 'start', bgColor: 'transparent', value: '' },
```
am I right that `firstSelectedDay` is always `position: 'start'` when `secondSelectedDay` is always `position: 'end'`?

OlehFilyak replied:
Yes, you are right. I'll change it according to what you wrote above.

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -32,18 +35,100 @@ interface IMaterialsDate {
+  const handleDatePicker = (option: string, value: Date | null) => {
+    if (value) {
+      const changedValue = value.setHours(23, 59, 59, 0).toString();
+      if (
```
I'm a bit lost here. Could you please add some short description why do we check what we check here?

OlehFilyak replied:
Okay. I'll try. Value can be Date, or null. In this line:
                onChange={(value) => {
                  handleDatePicker('start', value);
                }}
                We had an error
                Argument of type 'MaterialUiPickersDate' is not assignable to parameter of type 'Date'.
  Type 'null' is not assignable to type 'Date'.

skyboyer replied:
well, this explains `if (value)` part but not `option ==='end' && changedValue != ....`

also I'm not in the context why `setChanges` gets just a single value. does the parent element also have "start + end dates"? does it make sense to rely just on the state in parent instead of having 2 different sources of the truth?

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -54,39 +139,49 @@ export const AdminDatePicker: React.FC<IMaterialsDate> = ({
+              <DateRangeOutlinedIcon />
+            ) : (
+              <div>
+                {t(langTokens.admin.from)} {parsDate(start)}{' '}
```
I strongly suggest to code this as a single translation like

```
З {{parsedStartDate}} по {{parsedEndDate}}
```

Something like https://github.com/ita-social-projects/dokazovi-fe/blob/develop/src/locales/uk/parts/admin.ts#L79

You see, in different languages structure may be different, e.g. "from" and "to" might be combined into single word between dates or maybe grammar rules requires us to add text _after_ dates, not between. Having this coded in single solid message template gives much more flexibility with no disadvantages.

OlehFilyak committed:
bugFix(AdminDatePicker)

change the structure of the reducer
add types for ISelectedDayType
change langTokens

skyboyer commented on src/locales/uk/parts/admin.ts:
```diff
@@ -46,8 +46,7 @@ export const admin = {
   changeDateSuccess: 'Дату публікації змінено',
   scheduleDateSuccess: 'Дату публікації заплановано',
   schedulePublishTitle: 'Запланувати публікацію?',
-  from: 'з',
```
it's not used anywhere else, is it?

OlehFilyak replied:
No, this code isn't used anywhere. I searched in the project before I removed this line.

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -32,18 +35,93 @@ interface IMaterialsDate {
+    firstSelectedDay: ISelectedDayType;
+    secondSelectedDay: ISelectedDayType;
+  },
+  action: { type: string; payload: ISelectedDayType },
```
`type` is better to be typed as [a discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions)

OlehFilyak replied:
Sorry, I didn't understand you before, but now I changed it.

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -54,39 +132,51 @@ export const AdminDatePicker: React.FC<IMaterialsDate> = ({
+                  handleDatePicker('start', value);
+                }}
+                format="dd-mm-yyyy"
+                maxDate={today}
```
how does code behave if we select `start` > `end`?

OlehFilyak replied:
We have one handleDatePicker and when we send two different flags - 'start', and 'end'. When we click on the first DatePicker we automatically send 'start', and when we click on the second DatePicker we automatically send 'end'. This logic was written before I started rewriting this component.

skyboyer replied:
let's me explain what I mean

say, today is 06/09

i'm setting end date through second picker to 31/08
now I'm setting start date through first picker to 02/09

then start date> end date

how will it behave in this case?

skyboyer replied:
so how does it work in such a case?

OlehFilyak replied:
I changed the logic in this component. And when you set the end date through the second picker to 31/08, now you can't set the start date through the first picker to 02/09. The start day must be earlier than the end day.

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -54,39 +132,51 @@ export const AdminDatePicker: React.FC<IMaterialsDate> = ({
-              format="dd-mm-yyyy"
-              minDate={start}
-            />
+            <Box className={datepickerClassesStart.datePicker}>
```
I think, instead of putting color into a reducer and then create 2 different `datepickerClassesStart` and `datepickerClassesEnd` we could instead:

```
  const datepickerClassesSelectedDate = useStylesForDatePicker({
    bgColorForDatePicker: 'grey',
  });
  const datepickerClassesNoSelection = useStylesForDatePicker({
    bgColorForDatePicker: 'transparent',
  });
......
<Box className={start ? datepickerClassesSelectedDate.datePicker :  datepickerClassesNoSelection.datePicker }>
....
<Box className={end ? datepickerClassesSelectedDate.datePicker :  datepickerClassesNoSelection.datePicker }>
```

A bit repetitive? may be. but we don't have any other difference between Start Date and End Date selector, we just need to suppress current day highlighted, right?

Also this would allow us to get rid of extra state(`useReducer` it's also a state) and consume `end`/`start` again.

--------

Ideally, I'd like to have `DokazoviDatePicker.tsx` which would implement logic under the hood:

1. no highlighting for the current day if no selection
2. clicking currently selected day unselects it

This way we would re-use it across the app getting unified UX.

Do we use DatePicker elsewhere?

OlehFilyak replied:
We use DatePicker in SetPublicationDateModal. But not use AdminDatePicker.

skyboyer replied:
do we expect current day to be highlighted in SetPublicationDateModal?

OlehFilyak replied:
No, I asked Denis and he said that we need there the same behavior as this.

skyboyer replied:
understood. Then I propose to extract this into a separate component - also `<DatePicker` but which will have all this extra styling. All extra props should be passed down to `material-ui/pickers`' DatePicker. Then replacing it will just a change in `import .. from ` line

Easier to reuse, no risks of inconsistent changes

skyboyer replied:
so will you extract this into a separate component, so reducer here will not orchestrate colors?

OlehFilyak replied:
Sorry, I didn't write it. I tried to do that, but if we move in this direction, we greatly increase the duplication and confusion of the code, because style management is closely related to value in this component. I tried to rewrite the component without extracting this logic. It's a specific component, and I don't think we should extract it in a separate file. You can look at my attempt to rewrite the component in a zip file.
[AdminDatePicker 2.tsx.zip](https://github.com/ita-social-projects/dokazovi-fe/files/12584286/AdminDatePicker.2.tsx.zip)

skyboyer replied:
this is what I thought about:

```
import { DatePicker as MuiDatePicker } from '@material-ui/pickers';
import {createMuiTheme, makeStyles} from '@material-ui/core';

const useStyles = makeStyles((theme) => ({
  container: {
    display: 'contents',
    '& .MuiPickersDay-daySelected': {
      backgroundColor: ({ isHighlightingEnabled }) => isHighlightingEnabled ? theme.palette.primary.main : 'transparent',
      color: ({ isHighlightingEnabled }) => isHighlightingEnabled ? theme.palette.primary.contrastText : theme.palette.text.primary
    }
  }
}))

export function DatePicker(props) {
  const isHighlightingEnabled = !!props.value;
  const classes = useStyles({ isHighlightingEnabled })
  const onChange = useCallback((value) => {
    if (String(value) === String(props.value)) {
      props.onChange(null)
    } else {
      props.onChange(value)
    }
  }, [props.value, props.onChange])

  return (<div className={classes.container}>
    <MuiDatePicker 
      {...props} 
      variant="static"
      onChange={onChange} 
    />
  </div>)
}
```

and then instead of `import { DatePicker } from '@material-ui/pickers'` we could `import { DatePicker } from '../ourCustomDatePicker.tsx'` and use it just like if `material-ui`'s version would patched and worked with "unsetting" date and suppressing highlighting today day if `value` is `null`

the big disadvantage of that method - styles will work fine only if `variant="static"`, otherwise Dialog is rendered outside of our "extra div with custom class" and override does not work

What do you say?

skyboyer commented on src/views/Profile/AdminTable/AdminDatePicker.tsx:
```diff
@@ -32,18 +35,93 @@ interface IMaterialsDate {
+
+  const handleDatePicker = (option: string, value: Date | null) => {
+    if (value) {
+      const changedValue = value.setHours(23, 59, 59, 0).toString();
```
just to understand better: why to set to `23:59:59` not to `00:00:00`?

OlehFilyak replied:
We need to set 'setHours' for the correct comparison without different hours. If we set 00:00:00 it'll automatically transfer us to the previous day and apply logic for it.

skyboyer replied:
sorry, I don't get it :) which code exactly threats "2023-09-06 00:00:00" as September, 5th?

OlehFilyak replied:
If we write value.setHours(0, 0, 0, 0).toString() and choose 2023-09-06, we take hilghlighting for 2023-09-05
<img width="881" alt="Знімок екрана 2023-09-06 о 15 43 41" src="https://github.com/ita-social-projects/dokazovi-fe/assets/73895897/b017b8d7-39a0-4e4f-9ea2-0071115234ea">

skyboyer replied:
I see. I suspect depending on current time zone, setting date to 23:59:59 may instead highlight next day
May you check with changing local timezone e.g. to USA time?

skyboyer commented on src/views/Profile/AdminTable/styles/AdminDatePicker.styles.ts:
```diff
@@ -13,3 +13,12 @@ export const useStyles = makeStyles((theme: Theme) => ({
+
+export const useStylesForDatePicker = makeStyles({
+  datePicker: {
+    '& .MuiPickersDay-daySelected': {
```
btw btw, I see `disableHighlightToday` for [docs on](https://mui.com/x/api/date-pickers/date-picker/#DatePicker-prop-disableHighlightToday) newer version of material-ui. And looking for this property over legacy 3.3.10 I see [some usage](https://github.com/search?q=repo%3Amui%2Fmaterial-ui-pickers%20disableHighlightToday&type=code). May it be that it worked also in 3.3.10?

skyboyer replied:
I've checked in codesandbox and surprisingly it does not allow me to add `@material-ui/pickers@3.3.10` like it if was unsupported. So I'm asking you to check this on Dokazovi,

OlehFilyak replied:
I've already checked it, and it doesn't allow me to add it too.

OlehFilyak replied:
We can't add "disableHighlightToday", because we don't have this property in our library version. I find it in node modules.

skyboyer replied:
> I find it in node modules.

I guess, "I did not find", right?

OlehFilyak committed:
bugFix(AdminDatePicker)

add correct types
change logic for definition changedValue
change value for maxDate on first DatePicker
